daemon/src/connection.rs

```diff
@@ -116,9 +116,13 @@ impl State {
             return false;
         }
 
+        let heart_beat = self
+            .last_heartbeat()
+            .map(|heart_beat| format!("{heart_beat}"))
+            .unwrap_or_else(|| "None".to_string());
+        let seconds_since_heart_beat = duration_since_last_heartbeat.as_secs();
         tracing::warn!(
-            "Disconnecting due to lack of heartbeat. Last heartbeat: {:?}",
-            self.last_heartbeat()
+            "Disconnecting due to lack of heartbeat. Last heartbeat was {seconds_since_heart_beat} seconds ago. At {heart_beat}",
```

Comment:
Perhaps using tracing fields for the duration and the timestamp would be even better?

I find multi-sentence log messages to always be a bit of a code smell.